### PR TITLE
AQE support.

### PR DESCRIPTION
**Pull request by FMX**
# [FEATURE] AQE support.

### What changes were proposed in this pull request?
To support spark AQE.

AQE支持

### Related issues.
#57 

### How was this patch tested?
On cluster test.

/cc @wangshengjie123 

/assign @waitinfuture

### CODE REVIEW

File: common/src/main/java/com/aliyun/emr/rss/common/network/server/ShuffleFileSortTask.java [hunk 1]
```diff
@@ -0,0 +1,291 @@
+/*
```
[the remark this answers is not in the document]
Comment (waitinfuture) [reply common/src/main/java/com/aliyun/emr/rss/common/network/server/ShuffleFileSortTask.java:1]: Rename to PartitionFIleSorter

File: common/src/main/java/com/aliyun/emr/rss/common/network/server/ShuffleFileSortTask.java [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    return this.originFileLen < this.maxInMemSortSize;
+  }
+
+  class ShuffleBlockInfo {
+    protected int offset;
+    protected int length;
+  }
+
+  public interface SortCompleteCallback {
+    void onComplete();
```
Comment (waitinfuture) [inline common/src/main/java/com/aliyun/emr/rss/common/network/server/ShuffleFileSortTask.java:289]: onSuccess, onFailure

File: common/src/main/proto/TransportMessages.proto [hunk 3]
```diff
@@ -14,7 +14,7 @@ enum MessageType {
   RELEASE_SLOTS_RESPONSE = 8;
   REQUEST_SLOTS_RESPONSE = 9;
   REVIVE = 10;
-  REVIVE_RESPONSE = 11;
+  LOCATION_RENEWAL_RESPONSE = 11;
```
Comment (waitinfuture) [inline common/src/main/proto/TransportMessages.proto:17]: CHANGE_LOCATION_RESPONSE

File: common/src/main/proto/TransportMessages.proto [hunk 4]
```diff
@@ -154,11 +157,23 @@ message PbRevive {
   int32 epoch = 6;
   PbPartitionLocation oldPartition = 7;
   int32 status = 8;
+  int64 splitThreshold = 9;
```
Comment (waitinfuture) [inline common/src/main/proto/TransportMessages.proto:160]: remote and other places

File: common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala [hunk 5]
```diff
@@ -622,7 +657,7 @@ object ControlMessages extends Logging{
 
   def fromTransportMessage(message: TransportMessage): Message = {
     message.getType match {
-      case UNKNOWN_MESSAGE =>
+      case UNKNOWN_MESSAGE | UNRECOGNIZED =>
```
Comment (waitinfuture) [inline common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala:660]: why check UNRECOGNIZED

File: README.md [hunk 6]
```diff
@@ -182,10 +182,9 @@ spark.rss.shuffle.writer.mode hash
 # we recommend set spark.rss.push.data.replicate to true to enable server-side data replication 
 spark.rss.push.data.replicate true
 
-# Note: RSS didn`t support Spark AQE now, but we`ll support it soon.
-spark.sql.adaptive.enabled false
+# we recommend set localShuffleReader to false to get better performance of RSS
```
Comment (waitinfuture) [inline README.md:185]: give recommended AQE configs

File: client-spark/shuffle-manager-2/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java [hunk 7]
```diff
@@ -90,9 +92,8 @@
   private byte[][] sendBuffers;
   private int[] sendOffsets;
 
-  private final long[] mapStatusLengths;
+  private final LongAdder[] mapStatusLengths;
```
[the remark this answers is not in the document]
Comment (FMX) [reply client-spark/shuffle-manager-2/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java:95]: SortBased Pusher has already changed.

File: client/src/main/java/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 8]
```diff
@@ -906,7 +979,7 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
           })
         } else {
           // destroy success buffers
-          val destroyAfterRetry = retrySlots.asScala.filterKeys(!failedAfterRetry.contains(_))
+          val destroyAfterRetry = retrySlots.asScala.filterKeys(!failedAfterRetry.contains(_)).toMap
```
[the remark this answers is not in the document]
Comment (FMX) [reply client/src/main/java/com/aliyun/emr/rss/client/write/LifecycleManager.scala:982]: yes, it`s for scala 2.11.

File: client/src/main/java/com/aliyun/emr/rss/client/ShuffleClientImpl.java [hunk 9]
```diff
@@ -555,6 +589,59 @@ public void onFailure(Throwable e) {
...
+          return;
+        }
+        reducerSplittingSet.add(reduceId);
+      }
+
+      try {
+        ConcurrentHashMap<Integer, PartitionLocation>
+          currentShuffleLocs = reducePartitionMap.get(shuffleId);
+        LocationRenewalResponse shuffleSplitResponse =
+          driverRssMetaService.askSync(
```
[the remark this answers is not in the document]
Comment (FMX) [reply client/src/main/java/com/aliyun/emr/rss/client/ShuffleClientImpl.java:610]: Scala future cannot be converted to java future until Scala 2.13.

File: common/src/main/java/com/aliyun/emr/rss/common/protocol/message/StatusCode.java [hunk 10]
```diff
@@ -43,7 +43,10 @@
   PushDataFailNonCriticalCause(17),
   PushDataFailSlave(18),
   PushDataFailMain(19),
-  PushDataFailPartitionNotFound(20);
+  PushDataFailPartitionNotFound(20),
+
+  HardSplit(21),
+  SortSplit(22);
```
Comment (waitinfuture) [inline common/src/main/java/com/aliyun/emr/rss/common/protocol/message/StatusCode.java:49]: SoftSplit

File: common/src/main/java/com/aliyun/emr/rss/common/protocol/ShuffleSplitMode.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.aliyun.emr.rss.common.protocol;
+
+public enum ShuffleSplitMode {
+  // nonstrict means shuffle file reach split threshold and will receive data until shuffle split
+  // request complete.
+  nonstrict(0),
```
Comment (waitinfuture) [inline common/src/main/java/com/aliyun/emr/rss/common/protocol/ShuffleSplitMode.java:23]: soft

File: common/src/main/java/com/aliyun/emr/rss/common/protocol/ShuffleSplitMode.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  // strict means shuffle file reach split threshold and will stop receive data
+  strict(1);
```
Comment (waitinfuture) [inline common/src/main/java/com/aliyun/emr/rss/common/protocol/ShuffleSplitMode.java:25]: hard

File: common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala [hunk 13]
```diff
@@ -721,12 +718,44 @@ object RssConf extends Logging {
...
 
-  def supportAdaptiveQueryExecution(conf: RssConf): Boolean = {
-    conf.getBoolean("rss.support.adaptiveQueryExecution", false)
+  def shuffleSplitThreshold(conf: RssConf): Long = {
+    val threshold = conf.getSizeAsBytes("rss.shuffle.split.threshold", "256m")
+    Math.min(threshold, 1717986918)
+  }
+
+  def shuffleSplitMode(conf: RssConf): ShuffleSplitMode = {
+    val modeStr = conf.get("rss.shuffle.split.mode", "tolerant")
```
Comment (waitinfuture) [inline common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala:727]: unify to soft,hard

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFileSorter.java [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+        throw new IOException("Unexpected EOF, file name : " + originFile.getAbsolutePath());
+      }
+    }
+  }
+
+  private int transferFully(FileChannel originChannel, FileChannel targetChannel,
+    int offset, int length) throws IOException {
+    int transferedSize = 0;
+    while (transferedSize != length) {
+      transferedSize += originChannel.transferTo(offset, length, targetChannel);
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFileSorter.java:287]: should update offset and length

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala [hunk 15]
```diff
@@ -836,16 +889,91 @@ private[deploy] class Worker(
...
+          logInfo(s"[Worker] sorting file ${shuffleFileId}")
+          shuffleBoundSortingShuffleFiles.add(shuffleFileId)
+          val sortSuccess = new AtomicBoolean(false)
+
+          val sortTask = new PartitionFileSorter(fileWriter.getFile, readChunkSize,
+            sortMaxSingleFileRatio, fileWriter.getFileLength, new SortCallback {
+
+              override def onSuccess(): Unit = {
+                sortOver()
+                sortSuccess.set(false)
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala:932]: why set false

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala [hunk 16]
```diff
@@ -836,16 +889,91 @@ private[deploy] class Worker(
...
-    new FileInfo(fileWriter.getFile, fileWriter.getChunkOffsets, fileWriter.getFileLength)
+
+    if (endMapIndex != Int.MaxValue) {
+      val shuffleFileId = shuffleKey + fileName;
+      val shuffleBoundSortedShuffleFiles = sortedShuffleFiles.compute(shuffleKey,
+        newOpenStreamMapFunc)
+      val shuffleBoundSortingShuffleFiles = sortingShuffleFiles.compute(shuffleKey,
+        newOpenStreamMapFunc)
+
+      if (!shuffleBoundSortedShuffleFiles.contains(shuffleFileId)) {
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala:921]: synchronization issues here

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala [hunk 17]
```diff
@@ -836,16 +889,91 @@ private[deploy] class Worker(
...
+      s" startMapIndex ${startMapIndex} endMapIndex ${endMapIndex}")
     // find FileWriter responsible for the data
     val fileWriter = localStorageManager.getWriter(shuffleKey, fileName)
     if (fileWriter eq null) {
       logWarning(s"File $fileName for $shuffleKey was not found!")
       return null
     }
-    new FileInfo(fileWriter.getFile, fileWriter.getChunkOffsets, fileWriter.getFileLength)
+
+    if (endMapIndex != Int.MaxValue) {
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala:914]: Move all partition file sort code from Worker into PartitionFilesSorter

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala [hunk 18]
```diff
@@ -144,6 +137,48 @@ private[deploy] class Worker(
...
+          while (!memoryTracker.sortMemoryReady()) {
+            Thread.sleep(20)
+          }
+          shuffleSortExecutors.submit(new Runnable {
+            override def run(): Unit = {
+              task.sort()
+              if (task.inMemSort()) {
+                memoryTracker.releaseSortMemory(task.getOriginFileLen)
+              } else {
+                memoryTracker.reserveSortMemory(1024 * 1024)
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala:168]: release

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala [hunk 19]
```diff
@@ -144,6 +137,48 @@ private[deploy] class Worker(
...
+
+  private val shuffleSortSchedulerThread = new Thread(new Runnable {
+    override def run(): Unit = {
+      try {
+        while (true) {
+          val task = shuffleSortTaskDeque.take();
+          if (task.inMemSort()) {
+            memoryTracker.reserveSortMemory(task.getOriginFileLen)
+          } else {
+            memoryTracker.reserveSortMemory(1024 * 1024)
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala:157]: 1024 * 1024 should be configurable

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala [hunk 20]
```diff
@@ -836,16 +889,91 @@ private[deploy] class Worker(
...
+      val shuffleBoundSortingShuffleFiles = sortingShuffleFiles.compute(shuffleKey,
+        newOpenStreamMapFunc)
+
+      if (!shuffleBoundSortedShuffleFiles.contains(shuffleFileId)) {
+        if (!shuffleBoundSortingShuffleFiles.contains(shuffleFileId)) {
+          logInfo(s"[Worker] sorting file ${shuffleFileId}")
+          shuffleBoundSortingShuffleFiles.add(shuffleFileId)
+          val sortSuccess = new AtomicBoolean(false)
+
+          val sortTask = new PartitionFileSorter(fileWriter.getFile, readChunkSize,
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala:927]: There should be one PartitionFilesSorter instance in Worker, and refine apis of it

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFileSorter.java [hunk 21]
```diff
@@ -0,0 +1,301 @@
+/*
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFileSorter.java:1]: Rename to PartitionFilesSorter

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFileSorter.java [hunk 22]
```diff
@@ -0,0 +1,301 @@
...
+      logger.error("Shuffle sort file failed , detail :", e);
+      sortCallBack.onFailure();
+    } finally {
+      synchronized (this) {
+        notify();
+      }
+    }
+  }
+
+  public FileInfo resolve(int startMapIndex, int endMapIndex) {
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFileSorter.java:164]: cache in memory in stead of read from file

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFileSorter.java [hunk 23]
```diff
@@ -0,0 +1,301 @@
...
+        info.length = length;
+        blockInfos.add(info);
+      }
+      indexMap.put(mapId, blockInfos);
+    }
+    cleanBuffer(indexBuf);
+    return indexMap;
+  }
+
+  private ArrayList<Long> refreshChunkOffsets(int startMapIndex, int endMapIndex,
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFileSorter.java:230]: getChunkOffsets

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java [hunk 24]
```diff
@@ -0,0 +1,500 @@
...
+      FileSorter fileSorter = new FileSorter(fileWriter.getFile(), fetchBlockSize,
+        fileWriter.getFileLength(), null, sortInMem, shuffleSortKey, shuffleKey);
+
+      synchronized (shuffleBoundSortedShuffleFiles) {
+        if (shuffleBoundSortedShuffleFiles.contains(shuffleSortKey)) {
+          fileInfo = fileSorter.resolve(startMaxIndex, endMapIndex);
+        }
+      }
+
+      synchronized (shuffleBoundSortingShuffleFiles) {
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java:141]: refactor here, refer to handleRevive

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java [hunk 25]
```diff
@@ -0,0 +1,500 @@
...
+        if (!shuffleBoundSortedShuffleFiles.contains(shuffleSortKey)) {
+          shuffleBoundSortedShuffleFiles.add(shuffleSortKey);
+          AtomicBoolean sortSuccess = new AtomicBoolean(false);
+          SortCallback callback = new SortCallback() {
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java:145]: remove the callback, sorted and sorting be updated in FileSorter.sort

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java [hunk 26]
```diff
@@ -0,0 +1,500 @@
...
+  class FileSorter {
+    private final File originFile;
+    private final String sortedFileName;
+    private final String indexFileName;
+    private final long chunkSize;
+    private final long originFileLen;
+    private final String shuffleFileId;
+    private final String shuffleKey;
+
+    private SortCallback sortCallBack;
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java:230]: remove

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java [hunk 27]
```diff
@@ -0,0 +1,500 @@
...
+public class PartitionFilesSorter {
+  private static Logger logger = LoggerFactory.getLogger(PartitionFilesSorter.class);
+  public static final String SORTED_SUFFIX = ".sorted";
+  public static final String INDEX_SUFFIX = ".index";
+
+  private final ConcurrentHashMap<String, Set<String>> sortedShuffleFiles =
+    new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, Set<String>> sortingShuffleFiles =
+    new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, Map<String, ByteBuffer>> cachedIndexes =
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java:57]: cache Map<Integer, List<ShuffleBlockInfo>>, return value of readIndex

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java [hunk 28]
```diff
@@ -0,0 +1,500 @@
...
+        List<ShuffleBlockInfo> list = entry.getValue();
+        indexBuf.putInt(mapId);
+        indexBuf.putInt(list.size());
+        list.forEach(info -> {
+          indexBuf.putInt(info.offset);
+          indexBuf.putInt(info.length);
+        });
+      }
+
+      Map<String, ByteBuffer> cacheMap = cachedIndexes.compute(shuffleKey, (k, v) -> {
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java:398]: computeIfAbsent

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java [hunk 29]
```diff
@@ -0,0 +1,435 @@
...
+
+        sortedShuffleFiles.get(shuffleKey).add(shuffleFileId);
+        sortingShuffleFiles.get(shuffleKey).remove(shuffleFileId);
+        sortComplete = true;
+      } catch (Exception e) {
+        exception = e;
+      }
+    }
+
+    public FileInfo resolve(int startMapIndex, int endMapIndex) {
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java:303]: put this method and dependent methods in outer class

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java [hunk 30]
```diff
@@ -0,0 +1,435 @@
...
+
+  public void cleanup(HashSet<String> expirtedShuffleKeys) {
+    for (String expirtedShuffleKey : expirtedShuffleKeys) {
+      sortingShuffleFiles.remove(expirtedShuffleKey);
+      sortedShuffleFiles.remove(expirtedShuffleKey);
+      cachedIndexMaps.remove(expirtedShuffleKey);
+    }
+  }
+
+  private void cleanBuffer(ByteBuffer buf) {
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java:170]: no need for this method

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java [hunk 31]
```diff
@@ -0,0 +1,435 @@
...
+
+      Set<String> sorted =
+        sortedShuffleFiles.computeIfAbsent(shuffleKey, v -> ConcurrentHashMap.newKeySet());
+      Set<String> sorting =
+        sortingShuffleFiles.computeIfAbsent(shuffleKey, v -> ConcurrentHashMap.newKeySet());
+
+      boolean sortInMem = fileWriter.getFileLength() < maxSingleFileInMemSize;
+
+      FileSorter fileSorter = new FileSorter(fileWriter.getFile(), fetchBlockSize,
+        fileWriter.getFileLength(), sortInMem, shuffleSortKey, shuffleKey);
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java:121]: shuffleSortKey -> fileId

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java [hunk 32]
```diff
@@ -0,0 +1,435 @@
...
+    private final long chunkSize;
+    private final long originFileLen;
+    private final String shuffleFileId;
+    private final String shuffleKey;
+    private final boolean inMemSort;
+    private boolean sortComplete;
+    private Exception exception;
+
+    FileSorter(File originFile, long chunkSize, long originFileLen, boolean inMemSort,
+      String shuffleFileId, String shuffleKey) {
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java:198]: shuffleFileId -> fileId

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java [hunk 33]
```diff
@@ -0,0 +1,435 @@
...
+        sortedFileChannel.close();
+        if (!originFile.delete()) {
+          logger.warn("clean origin file failed, origin file is : {}",
+            originFile.getAbsolutePath());
+        }
+        writeIndex(sortedBlockInfoMap);
+
+        sortedShuffleFiles.get(shuffleKey).add(shuffleFileId);
+        sortingShuffleFiles.get(shuffleKey).remove(shuffleFileId);
+        sortComplete = true;
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/PartitionFilesSorter.java:297]: two status: Success, Fail

File: common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala [hunk 34]
```diff
@@ -721,18 +718,54 @@ object RssConf extends Logging {
     conf.getDouble("rss.slots.usage.overload.percent", 0.95)
   }
 
-  def supportAdaptiveQueryExecution(conf: RssConf): Boolean = {
-    conf.getBoolean("rss.support.adaptiveQueryExecution", false)
+  def shuffleSplitThreshold(conf: RssConf): Long = {
+    val threshold = conf.getSizeAsBytes("rss.shuffle.split.threshold", "256m")
+    Math.min(threshold, 1717986918)
```
Comment (waitinfuture) [inline common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala:723]: delete this line

File: common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala [hunk 35]
```diff
@@ -721,18 +718,54 @@ object RssConf extends Logging {
...
+  }
+
+  def shuffleSplitMode(conf: RssConf): ShuffleSplitMode = {
```
Comment (waitinfuture) [inline common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala:726]: partitionSplitMode

File: common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala [hunk 36]
```diff
@@ -721,18 +718,54 @@ object RssConf extends Logging {
...
+    }
+  }
+
+  def shuffleClientSplitPoolSize(conf: RssConf): Int = {
+    conf.getInt("rss.shuffle.client.split.pool.size",
+      Math.max(8, Runtime.getRuntime().availableProcessors()))
+  }
+
+  def shuffleSortSchedulerSize(conf: RssConf): Int = {
+    conf.getInt("rss.shuffle.sort.scheduler.size", 4)
```
Comment (waitinfuture) [inline common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala:742]: 8

File: common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala [hunk 37]
```diff
@@ -721,18 +718,54 @@ object RssConf extends Logging {
...
+    conf.getTimeAsMs("rss.shuffle.sort.timeout", "220s")
   }
 
-  def trafficControlEnabled(conf: RssConf): Boolean = {
-    conf.getBoolean("rss.traffic.control.enabled", true)
+  def shuffleSortMaxMemoryRatio(conf: RssConf): Double = {
+    conf.getDouble("rss.shuffle.sort.memory.max.ratio", 0.5)
+  }
+
+  def shuffleSortSingleFileMaxRatio(conf: RssConf): Double = {
```
Comment (waitinfuture) [inline common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala:757]: delete this limitation

File: common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala [hunk 38]
```diff
@@ -721,18 +718,54 @@ object RssConf extends Logging {
...
+    conf.getDouble("rss.shuffle.sort.single.file.max.ratio", 0.3)
   }
 
   def workerOffheapMemoryCriticalRatio(conf: RssConf): Double = {
     conf.getDouble("rss.worker.offheap.memory.critical.ratio", 0.9)
   }
 
+  def workerOffheapSortReserveMemory(conf: RssConf): Long = {
```
Comment (waitinfuture) [inline common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala:765]: memoryForSortLargeFile

File: CONFIGURATION_GUIDE.md [hunk 39]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (waitinfuture) [inline CONFIGURATION_GUIDE.md:1]: delete empty line

File: CONFIGURATION_GUIDE.md [hunk 40]
```diff
@@ -188,7 +188,13 @@ So we should set `rss.worker.flush.queue.capacity=6553` and each RSS worker has
...
-| `rss.traffic.control.enabled` | true | bool | Flow control |
 | `rss.worker.offheap.memory.critical.ratio` | 0.9 | float | Worker direct memory usage critical level ratio |
 | `rss.worker.memory.check.interval` | 10 | int | Timeunit is millisecond |
 | `rss.worker.memory.report.interval` | 10s | String | Timeunit is second |
+| `rss.shuffle.split.threshold` | 256m | String | Shuffle file split size, max value is 1.6GB |
+| `rss.shuffle.split.mode` | soft | String | sort, the shuffle file size might be larger than split threshold ; hard, the shuffle file size will be limited to split threshold  |
+| `rss.shuffle.client.split.pool.size` | 8 | int | Thread number to process shuffle split request in shuffle client. |
+| `rss.shuffle.sort.timeout` | 220 | int | Timeout for a shuffle file to sort |
+| `rss.shuffle.sort.memory.max.ratio` | 0.5 | double | Max ratio of sort memory |
+| `rss.shuffle.sort.single.file.max.ratio` | 0.3 | double | Max ratio of single shuffle file to sort in memory. If a shuffle file is larger than limit, it will be sorted on disk. |
```
Comment (waitinfuture) [inline CONFIGURATION_GUIDE.md:199]: delete

File: client-spark/shuffle-manager-3/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleManager.scala [hunk 41]
```diff
@@ -118,24 +120,22 @@ class RssShuffleManager(conf: SparkConf) extends ShuffleManager with Logging {
...
         } else {
           throw new UnsupportedOperationException(
-            s"Unrecognized shuffle write mode! ${RssConf.shuffleWriterMode(essConf)}")
+            s"Unrecognized shuffle write mode! ${RssConf.shuffleWriterMode(rssConf)}")
         }
       case _ => sortShuffleManager.getWriter(handle, mapId, context, metrics)
     }
   }
 
+
```
Comment (waitinfuture) [inline client-spark/shuffle-manager-3/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleManager.scala:137]: delete line

File: client-spark/shuffle-manager-3/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleReader.scala [hunk 42]
```diff
@@ -44,20 +47,21 @@ class RssShuffleReader[K, C](
...
-        readMetrics.incFetchWaitTime(time)
+        metrics.incFetchWaitTime(time)
     }
 
     val recordIter = (startPartition until endPartition).map(reduceId => {
       if (handle.numMappers > 0) {
         val start = System.currentTimeMillis()
-        val inputStream = essShuffleClient.readPartition(
-          handle.newAppId, handle.shuffleId, reduceId, context.attemptNumber())
+        val inputStream = essShuffleClient.readPartition(handle.newAppId, handle.shuffleId,
```
Comment (waitinfuture) [inline client-spark/shuffle-manager-3/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleReader.scala:63]: rssShuffleClient

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/ChunkFetchRpcHandler.java [hunk 43]
```diff
@@ -64,11 +64,16 @@ private String readString(ByteBuffer buffer) {
...
+    int endMapIndex = message.getInt();
 
     // metrics start
     source.startTimer(WorkerSource.FetchChunkTime(), shuffleKey);
-    FileInfo fileInfo = handler.handleOpenStream(shuffleKey, fileName);
+    FileInfo fileInfo = handler.handleOpenStream(shuffleKey, fileName, startMapIndex, endMapIndex);
+
     if (fileInfo != null) {
+      logger.debug("Received chunk fetch request {} {} {} {} get file info {}", shuffleKey,
+        fileName, startMapIndex, endMapIndex, fileInfo.toString());
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/ChunkFetchRpcHandler.java:76]: delete toString

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala [hunk 44]
```diff
@@ -74,12 +67,14 @@ private[deploy] class Worker(
...
+    workerDirectMemoryPressureCheckIntervalMs(conf), workerDirectMemoryReportIntervalSecond(conf),
+    shuffleSortMaxMemoryRatio(conf))
   private val localStorageManager = new LocalStorageManager(conf, workerSource, this)
-  if (RssConf.trafficControlEnabled(conf)) {
-    val memoryTracker = MemoryTracker.initialize(workerOffheapMemoryCriticalRatio(conf),
-      workerDirectMemoryPressureCheckIntervalMs(conf), workerDirectMemoryReportIntervalSecond(conf))
-    memoryTracker.registerMemoryListener(localStorageManager)
-  }
+  memoryTracker.registerMemoryListener(localStorageManager)
+  private val partitionsSorter = new PartitionFilesSorter(memoryTracker,
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala:75]: one parameter per line

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala [hunk 45]
```diff
@@ -74,12 +67,14 @@ private[deploy] class Worker(
     source
   }
 
+  val memoryTracker = MemoryTracker.initialize(workerOffheapMemoryCriticalRatio(conf),
+    workerDirectMemoryPressureCheckIntervalMs(conf), workerDirectMemoryReportIntervalSecond(conf),
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala:71]: one parameter per line

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala [hunk 46]
```diff
@@ -836,16 +849,28 @@ private[deploy] class Worker(
     }
   }
 
-  override def handleOpenStream(
-      shuffleKey: String, fileName: String): FileInfo = {
-    logDebug(s"Open file $fileName for $shuffleKey")
+  val newOpenStreamMapFunc =
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala:852]: unused

**Comment by waitinfuture** [comment]
LGTM. Thanks.

**Pull request merged**